### ISSUE TITLE
crash from gossiper::do_on_change_notifications when migration_manager stops on shutdown

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#5628]
Seen in https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/dtest-byo-parallel/113/artifact/logs-release.2.9/1579681759929_repair_additional_test.RepairAdditionalTest.repair_disjoint_data_test/node3.log
Build [451](https://jenkins.scylladb.com/view/master/job/scylla-master/job/build/451/artifact/scylla/build/release/)
Scylla version a74a82d4d245d75311c19d0d8fba404d5d75937e
```
INFO  2020-01-22 08:28:18,388 [shard 0] migration_manager - stopping migration service
INFO  2020-01-22 08:28:18,388 [shard 1] migration_manager - stopping migration service
INFO  2020-01-22 08:28:18,388 [shard 0] init - Shutting down migration manager was successful
INFO  2020-01-22 08:28:18,388 [shard 0] init - Shutting down database
INFO  2020-01-22 08:28:18,388 [shard 0] compaction_manager - Asked to stop
INFO  2020-01-22 08:28:18,388 [shard 0] compaction_manager - Stopped
INFO  2020-01-22 08:28:18,388 [shard 1] compaction_manager - Asked to stop
INFO  2020-01-22 08:28:18,388 [shard 1] compaction_manager - Stopped
INFO  2020-01-22 08:28:18,447 [shard 0] storage_service - Node 127.0.30.2 state jump to normal
scylla: ./seastar/include/seastar/core/sharded.hh:712: Service& seastar::sharded<T>::local() [with Service = service::migration_manager]: Assertion `local_is_initialized()' failed.
Aborting on shard 0.
Backtrace:
  0x0000000002a516a2
  0x00000000029f821b
  0x00000000029f8515
  0x00000000029f85b0
  0x00007f4c712cab1f
  /tmp/jenkins-3901ad4e/workspace/scylla-staging/fruch/dtest-byo-parallel/scylla/.ccm/scylla-repository/unstable/master/202001202212/scylla//opt/scylladb/libreloc/libc.so.6+0x000000000003c624
  /tmp/jenkins-3901ad4e/workspace/scylla-staging/fruch/dtest-byo-parallel/scylla/.ccm/scylla-repository/unstable/master/202001202212/scylla//opt/scylladb/libreloc/libc.so.6+0x00000000000258d8
  /tmp/jenkins-3901ad4e/workspace/scylla-staging/fruch/dtest-byo-parallel/scylla/.ccm/scylla-repository/unstable/master/202001202212/scylla//opt/scylladb/libreloc/libc.so.6+0x00000000000257a8
  /tmp/jenkins-3901ad4e/workspace/scylla-staging/fruch/dtest-byo-parallel/scylla/.ccm/scylla-repository/unstable/master/202001202212/scylla//opt/scylladb/libreloc/libc.so.6+0x0000000000034a65
  0x0000000001f3f8ff
  0x00000000019c2a53
  0x00000000019fc08f
  0x00000000019fcecb
  0x00000000019fd846
  0x000000000077294c
```

Decoded:
```
$ addr2line -Cfpie libexec/scylla 
  0x0000000001f3f8ff
  0x00000000019c2a53
  0x00000000019fc08f
  0x00000000019fcecb
  0x00000000019fd846
  0x000000000077294c
seastar::sharded<service::migration_manager>::local() at ././seastar/include/seastar/core/sharded.hh:712
 (inlined by) service::get_local_migration_manager() at ././service/migration_manager.hh:165
 (inlined by) service::storage_service::on_change(gms::inet_address, gms::application_state, gms::versioned_value const&) at ./service/storage_service.cc:1257
gms::gossiper::do_on_change_notifications(gms::inet_address, gms::application_state const&, gms::versioned_value const&) at ./gms/gossiper.cc:1630
 (inlined by) for_each<std::_List_iterator<seastar::shared_ptr<gms::i_endpoint_state_change_subscriber> >, gms::gossiper::do_on_change_notifications(gms::inet_address, const gms::application_state&, const gms::versioned_value&)::<lambda(auto:191&)> > at /usr/include/c++/9/bits/stl_algo.h:3876
 (inlined by) for_each<gms::gossiper::do_on_change_notifications(gms::inet_address, const gms::application_state&, const gms::versioned_value&)::<lambda(auto:191&)> > at ././gms/gossiper.hh:219
 (inlined by) gms::gossiper::do_on_change_notifications(gms::inet_address, gms::application_state const&, gms::versioned_value const&) at ./gms/gossiper.cc:1629
gms::gossiper::apply_new_states(gms::inet_address, gms::endpoint_state&, gms::endpoint_state const&) at ./gms/gossiper.cc:1587
 (inlined by) ~deferred_action at ././seastar/include/seastar/util/defer.hh:44
 (inlined by) gms::gossiper::apply_new_states(gms::inet_address, gms::endpoint_state&, gms::endpoint_state const&) at ./gms/gossiper.cc:1589
gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::{lambda(auto:1&&, auto:2&&)#2}::operator()<utils::chunked_vector<gms::inet_address, 131072ul>&, std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >&>(utils::chunked_vector<gms::inet_address, 131072ul>&, std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >&) const::{lambda(auto:1&&)#1}::operator()<gms::inet_address&>(utils::chunked_vector<gms::inet_address, 131072ul>&) const::{lambda()#1}::operator()()::{lambda()#1}::operator()() at ./gms/gossiper.cc:604
seastar::noncopyable_function<void ()>::direct_vtable_for<seastar::async<gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::{lambda(auto:1&&, auto:2&&)#2}::operator()<utils::chunked_vector<gms::inet_address, 131072ul>&, std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >&>(utils::chunked_vector<gms::inet_address, 131072ul>&, std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >&) const::{lambda(auto:1&&)#1}::operator()<gms::inet_address&>(utils::chunked_vector<gms::inet_address, 131072ul>&) const::{lambda()#1}::operator()()::{lambda()#1}>(seastar::thread_attributes, utils::chunked_vector<gms::inet_address, 131072ul>&, (std::result_of&&)...)::{lambda(seastar::async<seastar::future<>>(std::result_of<std::decay<auto:1>::type ()>, auto:1&&, seastar::futurize<std::result_of<std::decay<auto:1>::type ()>::type>)::work&)#1}::operator()(seastar::thread_attributes)::{lambda()#1}>::call(seastar::noncopyable_function<void ()> const*) at ././seastar/include/seastar/core/apply.hh:36
 (inlined by) apply<gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state>)::<lambda(auto:169&&, auto:170&&)> [with auto:169 = utils::chunked_vector<gms::inet_address>&; auto:170 = std::map<gms::inet_address, gms::endpoint_state>&]::<lambda(auto:171&&)> [with auto:171 = gms::inet_address&]::<lambda()> mutable::<lambda()> > at ././seastar/include/seastar/core/apply.hh:44
 (inlined by) apply_tuple<gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state>)::<lambda(auto:169&&, auto:170&&)> [with auto:169 = utils::chunked_vector<gms::inet_address>&; auto:170 = std::map<gms::inet_address, gms::endpoint_state>&]::<lambda(auto:171&&)> [with auto:171 = gms::inet_address&]::<lambda()> mutable::<lambda()> > at ././seastar/include/seastar/core/future.hh:1599
 (inlined by) apply<gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state>)::<lambda(auto:169&&, auto:170&&)> [with auto:169 = utils::chunked_vector<gms::inet_address>&; auto:170 = std::map<gms::inet_address, gms::endpoint_state>&]::<lambda(auto:171&&)> [with auto:171 = gms::inet_address&]::<lambda()> mutable::<lambda()> > at ././seastar/include/seastar/core/future.hh:1633
 (inlined by) operator() at ././seastar/include/seastar/core/thread.hh:265
 (inlined by) call at ././seastar/include/seastar/util/noncopyable_function.hh:101
```

https://github.com/scylladb/scylla/blob/a74a82d4d245d75311c19d0d8fba404d5d75937e/gms/gossiper.cc#L1568-L1589

**Comment by bhalevy** [comment]
@xemul said:
> in on_change in storage service it may want to pull new schema
> And this is the case
> We should either stop gossiper by that time, or keep migration manger alive

@bhalevy
> or fail gracefully if it's already stopped

@xemul
> Yes, but this looks like a bandaid for me :) it fails already, but … not extremely gracefully

**Comment by bhalevy** [comment]
Core dump [here](https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/dtest-byo-parallel/113/artifact/logs-release.2.10/1579683137703_repair_additional_test.RepairAdditionalTest.repair_disjoint_row_3nodes_test/node3-scylla.4851.1579683072.core.gz)
scylla-package [here](https://jenkins.scylladb.com/view/master/job/scylla-master/job/build/451/artifact/scylla/build/release/scylla-package.tar.gz)

**Comment by bhalevy** [comment]
Band-aid fix:
- Let’s try to check local_is_initialized before calling get_local_migration_manager in storage_service::on_change().
- Also, get the future (we’re running in a seastar::async context)

**Comment by bhalevy** [comment]
@scylladb/scylla-maint please backport 4839ca8 to all live branches (See #5765)

**Comment by tgrabiec** [comment]
Fixed in 4839ca849136841b4545931876b1345ee09affa1
